pkg/auth/oauth/registry/tokenauthenticator.go

```diff
@@ -52,20 +53,25 @@ func (a *TokenAuthenticator) AuthenticateToken(value string) (kuser.Info, bool,
 		return nil, false, fmt.Errorf("user.UID (%s) does not match token.userUID (%s)", u.UID, token.UserUID)
 	}
 
+	// Build group membership for the user
+	groupNames := sets.NewString()
+	// 1. Groups from the identity provider associated with the token
+	groupNames.Insert(token.IdentityProviderGroups...)
+	// 2. Groups directly in the user object (deprecated)
+	groupNames.Insert(u.Groups...)
+	// 3. Group API objects the user is listed as a member in
 	groups, err := a.groupMapper.GroupsFor(u.Name)
 	if err != nil {
 		return nil, false, err
 	}
-	groupNames := []string{}
 	for _, group := range groups {
-		groupNames = append(groupNames, group.Name)
+		groupNames.Insert(group.Name)
```

Comment:
Should this check if group with specific name isn't present in the set already? And insert it, only if not present yet -- something like in https://github.com/openshift/origin/blob/master/pkg/cmd/admin/groups/new.go#L127-L135